### PR TITLE
[Custom Image] Support tag for the images and global regions

### PR DESCRIPTION
**Pull request by Michaelvll**
Add an `image tag` for choosing a list of images provided by sky. Currently, we don't support specifying the image tag for Azure, as we previously did for `image_id`, mainly because Azure has more complex image setting.

The `image_id` still needs to be specified with the clouds, since we don't align the system version and the cuda version for different clouds at the moment.
Please refer to https://github.com/skypilot-org/skypilot-catalog/pull/3 for the details of the catalog.

Suggested method for using all AWS regions besides the US regions:
```
cd ~/.sky/catalog/v4
python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions

sky launch --cloud AWS
```

To launch a VM with Ubuntu 18.04
```
sky launch --cloud aws --image-id sky:gpu-ubuntu-1804
```

We also enforce the US regions will be tried first.

Tested:
- [x] `sky launch --cloud aws --image-id sky:gpu-ubuntu-1804 --gpus T4`
  candidate regions: `['us-west-2', 'us-east-2', 'us-east-1', 'us-west-1', 'eu-north-1', 'ca-central-1', 'eu-west-1', 'eu-west-2', 'eu-west-3', 'eu-south-1', 'me-south-1', 'ap-northeast-2', 'eu-central-1', 'ap-southeast-2', 'af-south-1', 'ap-northeast-1', 'ap-northeast-3', 'ap-southeast-1', 'ap-east-1']`
- [x] `python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions`
- [x] `sky launch --gpus A100:8 --cloud aws`: correctly failover through all the regions, starting from US regions.
- [x] `./tests/run_smoke_tests.sh test_minimal`
- [x] `sky launch --cloud gcp --zone europe-west4-a`
- [x] `sky launch --cloud azure --region australiaeast`
- [x] rendered locally

### CODE REVIEW

**Comment by concretevitamin** [comment]
We've seen demand for two filters:
- ubuntu: <version>
- cuda: <version> [less immediate]

Can this design support both filters?  Imagining an image catalog like
```
image_id,os,os_version,cuda_version,region
ami-1234,ubuntu,18.04,11.6,...
ami-5678,ubuntu,18.04,,...
```
etc.

**Comment by Michaelvll** [comment]
> We've seen demand for two filters:
> 
> * ubuntu: version
> * cuda: version [less immediate]
> 
> Can this design support both filters? Imagining an image catalog like
> 
> ```
> image_id,os,os_version,cuda_version,region
> ami-1234,ubuntu,18.04,11.6,...
> ami-5678,ubuntu,18.04,,...
> ```
> 
> etc.

Good point! I would prefer to use the tag (docker-style image selection) at least for now to unblock the user's use case in #1346, because:
1. We don't have a complete list of the images, e.g. we only have two versions of the system and a single version of the CUDA on AWS. CUDA versions can have a huge candidate list, and we may not want to let the users tune too many image-related options while the CUDA is backward compatible. To use filtering, we need to document the candidate lists for each filtering field to the user.
2. Tag is the nickname we assigned to groups of AMIs. The main purpose is to keep the image selection **coarse-grained** (we can later add functionality to filtering, but still support Tag as a convenient way to select the group of images).
3. Tag can re-use our current `image_id` field in our yaml instead of adding another field.

Things we need to do before using filtering:
1. Keep the default system(version) the same across the clouds.
2. Figure out a way to find the compatibility setting of the images across the cloud, which can be hard.
3. Figuring out the CUDA version for different images is non-trivial, which needs to actually launch the VM and check the CUDA version.

What do your think to leave the filtering-based (as well as CUDA selection) design to the future?

File: sky/clouds/service_catalog/common.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:49]: nit: `cloud` is used for printing a string name; can we pass `cloud_display_name: str`? Or use the dir name of `filename`.

File: sky/clouds/aws.py
Comment (concretevitamin) [inline sky/clouds/aws.py:225]: Make private? nit: self._get_image_id

File: sky/clouds/service_catalog/common.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:338]: Document that this can be None, e.g., when a region doesn't have an image that matches the tag.

File: sky/clouds/gcp.py
Comment (concretevitamin) [inline sky/clouds/gcp.py:220]: Is _IMAGE_ID_PREFIX still used?

File: sky/clouds/service_catalog/common.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:336]: Can't we hit this assert when region is None? E.g., many regions offer the tag `sky:k80-ubuntu-1804`?

File: sky/clouds/aws.py
Comment (concretevitamin) [inline sky/clouds/aws.py:104]: It seems this is not an internal logic error but can happen due to region input?

maybe: `raise ValueError('No image found in catalog for region {region_name}. Try setting a valid image_id.')`

nit: test `is None`.

File: sky/clouds/aws.py
Comment (concretevitamin) [inline sky/clouds/aws.py:119]: nit: image_id is None

File: sky/clouds/aws.py
Comment (concretevitamin) [inline sky/clouds/aws.py:123]: nit: No suitable image ..

I think requiring an implementor of `Cloud` to know about raising exceptions.ResourcesUnavailableError is info leakage, but I don't have a good suggestion right now to resolve it. Maybe add a TODO?

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:205]: outdated?

**Review by concretevitamin — COMMENTED** [review]
This looks great @Michaelvll. The reasoning of using this for now makes sense to me.

We may want a smoke test / an example to illustrate usage.

File: sky/clouds/service_catalog/aws_catalog.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:67]:
```suggestion
    # Hack: Enforce US regions are always tried first:
    #   [US regions sorted by price] + [non-US regions sorted by price]
```

File: sky/clouds/service_catalog/common.py
Comment (concretevitamin) [reply sky/clouds/service_catalog/common.py:343]: I'm missing something: what happens if we have
```
resources:
  cloud: aws
  image_id: sky:gpu-ubuntu-18.04
```
which will find multiple images with this tag, hitting this assert?

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:204]: L208-209 are a bit confusing now that L210 specify two timestamps. Not sure how to better phrase it however.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:253]: update return type?

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:275]:
```suggestion
                        help='Fetch all global regions, not just the U.S. ones.')
```

File: sky/clouds/service_catalog/gcp_catalog.py
Comment (concretevitamin) [inline sky/clouds/service_catalog/gcp_catalog.py:445]: Is this return type unintentionally changed?

**Review by concretevitamin — COMMENTED** [review]
Did a pass. Wdyt about changing `sky:` to `skypilot:`? Incrementally moving to a consistent name in code.

File: sky/clouds/service_catalog/aws_catalog.py
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py:67]: Done. Thanks!

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py:253]: Good catch! Fixed. Thanks!

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py:275]: Done. Thanks!

File: tests/test_smoke.py
Comment (Michaelvll) [reply tests/test_smoke.py:137]: It will correctly error out as following:
```
Task from YAML spec: examples/minimal.yaml
sky.exceptions.ResourcesMismatchError: Requested resources do not match the existing cluster.
  Requested:    1x AWS(None, image_id='sky:gpu-ubuntu-2004') 
  Existing:     1x AWS(m6i.2xlarge, image_id='sky:gpu-ubuntu-1804')
To fix: specify a new cluster name, or down the existing cluster first: sky down test-image
```

Added a check for it.

File: sky/clouds/service_catalog/common.py
Comment (Michaelvll) [reply sky/clouds/service_catalog/common.py:343]: It won't hit the assertion, since the method will only be called after the failover mechanism decides the region.

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks for the rapid PR and careful testing @Michaelvll!  Should we run all smoke tests  before merging?

**Comment by Michaelvll** [comment]
Yes, I will run the smoke test. Can we merge the PR in the service catalog, so that we can make sure the auto-downloading works as well? https://github.com/skypilot-org/skypilot-catalog/pull/3

File: docs/source/reference/faq.rst
Comment (concretevitamin) [inline docs/source/reference/faq.rst:79]:
```suggestion
(Advanced) How to make SkyPilot use all global regions?
```

File: docs/source/reference/faq.rst
Comment (concretevitamin) [inline docs/source/reference/faq.rst:82]:
```suggestion
By default, SkyPilot only supports the US regions on different clouds for convenience. If you want to utilize all global regions, please run the following command:
```

File: docs/source/reference/faq.rst
Comment (concretevitamin) [inline docs/source/reference/faq.rst:95]:
```suggestion
(the full name can be found in ``sky status``), and run the commands above.
```

**Review by concretevitamin — APPROVED** [review]
LGTM

File: docs/source/reference/faq.rst
Comment (Michaelvll) [reply docs/source/reference/faq.rst]: Done. Thanks!

File: docs/source/reference/faq.rst
Comment (Michaelvll) [reply docs/source/reference/faq.rst]: Done. Thanks!

File: docs/source/reference/faq.rst
Comment (Michaelvll) [reply docs/source/reference/faq.rst]: Done. Thanks!

File: docs/source/reference/faq.rst
Comment (Michaelvll) [reply docs/source/reference/faq.rst]: Done. Thanks!

File: sky/clouds/service_catalog/aws_catalog.py
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: renamed thanks!

**Comment by Michaelvll** [comment]
Passed smoke tests. Merging the catalog first, then merging this one.

**Pull request merged**
